Remove skipping task logic from Task Manager

ersin-erdal opened the pull request:
Towards: #176585

This PR removes the task skipping logic from TaskManager, PRs for Alerting and Actions will follow.

## To verify
Rules and actions should be still working as expected.

ersin-erdal commented:
/ci

ersin-erdal commented:
/ci

ersin-erdal commented:
/ci

ersin-erdal commented:
/ci

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

ymao1 reviewed (commented):
Verified by
* creating a rule on `main`
* modifying the `paramsSchema` on the task definition to add a new required field to the schema
* seeing the rule task get skipped up to the max times then run
* switching to this branch and removing the `requeue_invalid_tasks` config
* running the rule and seeing it run with no errors

Seems like everything is working as expected! 🎉 

One thing I noticed is that `numSkippedRuns` stays in the task doc in the task manager index. Do we want to keep that in the task doc since we're removing it from the task schema?

ersin-erdal commented:
> One thing I noticed is that `numSkippedRuns` stays in the task doc in the task manager index. Do we want to keep that in the task doc since we're removing it from the task schema?

I am also not sure, bulkUpdate partially updates the doc, so removing the field from post body is not enough.
We can write a migration but not sure if it's worth, tasks are not permanent docs.

ymao1 commented:
> We can write a migration but not sure if it's worth, tasks are not permanent docs.

That's true, I'd hate to have a migration for this, I'm just wondering if it might issues in the future if there's a field that could potentially be in the task document that's not in our schema. Task docs for recurring tasks do stick around for quite a while. cc @mikecote

mikecote commented:
>I'm just wondering if it might issues in the future if there's a field that could potentially be in the task document that's not in our schema.

I also feel this could be a potential issue down the road, especially as we add strict schema validation prior to saving. I feel our best bet is either a) leave it part of our TS and validation schemas or b) bump the model version with a migration removing the field.

ersin-erdal commented:
> > I'm just wondering if it might issues in the future if there's a field that could potentially be in the task document that's not in our schema.
> 
> I also feel this could be a potential issue down the road, especially as we add strict schema validation prior to saving. I feel our best bet is either a) leave it part of our TS and validation schemas or b) bump the model version with a migration removing the field.

Then I would go with migration. I'll add modelVersions with migration to TM. @ymao1 ?

ymao1 commented:
> > > I'm just wondering if it might issues in the future if there's a field that could potentially be in the task document that's not in our schema.
> > 
> > 
> > I also feel this could be a potential issue down the road, especially as we add strict schema validation prior to saving. I feel our best bet is either a) leave it part of our TS and validation schemas or b) bump the model version with a migration removing the field.
> 
> Then I would go with migration. I'll add modelVersions with migration to TM. @ymao1 ?

SGTM!

ersin-erdal commented:
I added modelVersions but looks like it doesn't support removing fields yet:
https://github.com/elastic/kibana/blob/53cc7414c430f83c1159e1fda2d4be1c7ce76dab/packages/core/saved-objects/core-saved-objects-server/model_versions.md#--mappings_deprecation

Therefore I added `numSkippedRuns` to deprecated mappings in modelVersions and to `TaskInstanceWithDeprecatedFields` with a deprecate annotation like we do for the deprecated `interval` field.

I created the taskSchema based on the TaskInstance interface. I used `schema.maybe` for the optional fields but some of them might be required indeed.

ymao1 commented on x-pack/plugins/task_manager/server/saved_objects/schemas/task.ts:
```diff
@@ -0,0 +1,39 @@
+  traceparent: schema.maybe(schema.string()),
+  user: schema.maybe(schema.string()),
+  scope: schema.maybe(schema.arrayOf(schema.string())),
+  ownerId: schema.nullable(schema.string()),
```
`ownerId` is defined as optional in the TaskInstance schema, so maybe this should be `schema.maybe(schema.nullable(schema.string()))`?

ymao1 replied:
Or maybe this is based off the `ConcreteTaskInstance` definition?

ersin-erdal replied:
Good catch, made me realise that I should use `SerializedConcreteTaskInstance` as it is used in the TaskStore.schedule of TM.

Updated the schema accordingly.

ymao1 replied:
It looks like `SerializedConcreteTaskInstance` inherits an `id` field from `ConcreteTaskInstance` that is a string, however, I don't see the `id` field in the actual task document?

ersin-erdal replied:
`id` is not saved in the doc, omited here: https://github.com/elastic/kibana/blob/main/x-pack/plugins/task_manager/server/task_store.ts#L553
 
but used as doc id in ES here: https://github.com/elastic/kibana/blob/main/x-pack/plugins/task_manager/server/task_store.ts#L171

ymao1 commented on x-pack/plugins/task_manager/server/saved_objects/index.ts:
```diff
@@ -72,5 +73,6 @@ export function setupSavedObjects(
         },
       } as estypes.QueryDslQueryContainer;
     },
+    modelVersions: taskModelVersions,
```
@pgayvallet These changes to the task manager model version were added after you approved the PR (for `serverless.yml` changes only). Could you take a look at these changes as well? Thanks!
